lnbits/extensions/boltz/boltz.py

```diff
@@ -0,0 +1,319 @@
+    - User generates preimage, RHASH. Sends RHASH to server.
+    - Server creates an LN invoice for RHASH.
+    - User pays LN invoice - except server needs to hold the HTLC as preimage is unknown.
+    - Server creates on-chain output locked to RHASH.
+    - User spends on-chain output, revealing preimage.
+    - Server fulfills HTLC using preimage.
+    Note: expected_onchain_amount_sat is BEFORE deducting the on-chain claim tx fee.
+    """
+
+    claim_privkey = ec.PrivateKey(os.urandom(32), True, net)
+    claim_pubkey_hex = hexlify(claim_privkey.sec()).decode("UTF-8")
+    preimage = os.urandom(32)
+    preimage_hash = sha256(preimage).hexdigest()
+
+    res = create_post_request(BOLTZ_URL + "/createswap", {
+        "type": "reversesubmarine",
+        "pairId": "BTC/BTC",
+        "orderSide": "buy",
+        "invoiceAmount": data.amount,
+        "preimageHash": preimage_hash,
+        "claimPublicKey": claim_pubkey_hex
+    })
+
+    swap = ReverseSubmarineSwap(
+        id = swap_id,
+        amount = data.amount,
+        wallet = data.wallet,
+        onchain_address = data.onchain_address,
+        instant_settlement = data.instant_settlement,
+        claim_privkey = claim_privkey.wif(net),
+        preimage = preimage.hex(),
+        boltz_id = res["id"],
+        invoice = res["invoice"],
+        timeout_block_height = res["timeoutBlockHeight"],
+        lockup_address = res["lockupAddress"],
+        onchain_amount = res["onchainAmount"],
+        redeem_script = res["redeemScript"],
+        time = getTimestamp(),
+    )
+
+    asyncio.ensure_future(wait_for_onchain_tx(swap))
+
+    # TODO: think about what happens if pay_invoice fails
+    # ensure_future is used because pay_invoice is stuck
+    # as long as boltz does not see the onchain tx
+    asyncio.ensure_future(pay_invoice(
+       wallet_id=data.wallet,
+       payment_request=res["invoice"],
+       description=f"reverse submarine swap for {data.amount} sats on boltz.exchange",
+       extra={"tag": "boltz"},
+    ))
+
+    return swap
+
+def listen_to_stream_swap_status(swap: ReverseSubmarineSwap):
+    messages = SSEClient(BOLTZ_URL + "/streamswapstatus?id=" + swap.boltz_id)
+    for msg in messages:
+        data = json.loads(msg.data)
+        if swap.instant_settlement == True:
+            status = "transaction.mempool"
+        else:
+            status = "transaction.confirmed"
+        if data["status"] == status:
+            return swap
+        if  data["status"] == "transaction.failed":
+            raise HTTPException(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR, detail="boltz status transaction.failed! no onchain funds at boltz"
+            )
+        if  data["status"] == "invoice.expired":
+            raise HTTPException(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR, detail="ln invoice has expired!"
+            )
+
+async def wait_for_onchain_tx(swap: Union[ReverseSubmarineSwap, SubmarineSwap]):
```

Comment:
This method should be made a little bit more resilient. Like reconnecting in case of a HTTP exception.
Also it would be good to start listening again to all swaps that don't have a successful or failed status on startup of lnbits.

Reply:
good idea, @arcbtc any tip on how to accomplish the startup checks?